File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/interpreter/lab/Unary.java
// Unary.java

import java.io.PrintWriter;

public abstract class Unary extends Operator
{
	public Unary(Expression aOperand1)
	{
		setOperand1(aOperand1);
	}

	public int getNumberOfUsesOfVariable(String aName)
	{
		int result = 0;

		if (getOperand1() == null)
		{
			throw new DpException("int Unary.getNumberOfUsesOfVariable(String aName) -- getOperand1() == null.");
		}
		result += getOperand1().getNumberOfUsesOfVariable(aName);

		return result;
	}

	public abstract void output(PrintWriter aPrintWriter);

	protected void outputPrefix(PrintWriter aPrintWriter, String aOperator)
	{
		aPrintWriter.print("(");
		aPrintWriter.print(aOperator);
		if (getOperand1() == null)
		{
			throw new DpException("void Unary.outputPrefix(PrintWriter aPrintWriter, String aOperator) -- getOperand1() == null.");
		}
		getOperand1().output(aPrintWriter);
		aPrintWriter.print(")");
	}

	protected void outputPostfix(PrintWriter aPrintWriter, String aOperator)
	{
		aPrintWriter.print("(");
		if (getOperand1() == null)
		{
			throw new DpException("void Unary.outputPostfix(PrintWriter aPrintWriter, String aOperator) -- getOperand1() == null.");
		}
		getOperand1().output(aPrintWriter);
		aPrintWriter.print(aOperator);
		aPrintWriter.print(")");
	}

	private void setOperand1(Expression aOperand1)
	{
		myOperand1 = aOperand1;
	}

	public Expression getOperand1()
	{
		return myOperand1;
	}

	private Expression myOperand1;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/interpreter/solution/unary.cpp
// unary.cpp

#include <iostream>
#include <string>
#include "dp_exception.h"
class expression;
#include "unary.h"

unary::unary(expression * a_operand1)
{
	set_operand1(a_operand1);
}

int unary::get_number_of_uses_of_variable(const std::string & a_name) const
{
	int result(0);

	if (get_operand1() == nullptr)
	{
		throw dp_exception("int unary::get_number_of_uses_of_variable(const std::string & a_name) const -- get_operand1() == nullptr.");
	}
	result += get_operand1()->get_number_of_uses_of_variable(a_name);

	return result;
}

void unary::output_prefix(std::ostream & a_ostream, const std::string & a_operator) const
{
	a_ostream << "(";
	a_ostream << a_operator;
	if (get_operand1() == nullptr)
	{
		throw dp_exception("void unary::output_prefix(std::ostream & a_ostream, const std::string & a_operator) const -- get_operand1() == nullptr.");
	}
	get_operand1()->output(a_ostream);
	a_ostream << ")";
}

void unary::output_postfix(std::ostream & a_ostream, const std::string & a_operator) const
{
	a_ostream << "(";
	if (get_operand1() == nullptr)
	{
		throw dp_exception("void unary::output_postfix(std::ostream & a_ostream, const std::string & a_operator) const -- get_operand1() == nullptr.");
	}
	get_operand1()->output(a_ostream);
	a_ostream << a_operator;
	a_ostream << ")";
}

void unary::set_operand1(expression * a_operand1)
{
	my_operand1 = a_operand1;
}

expression * unary::get_operand1() const
{
	return my_operand1;
}
